### PR TITLE
feat: allow initialising project with custom package name with `init --package-name`

### LINKED ISSUE #1269: Support for bundleid/packagename when initializing a new app
## Describe the Feature
Add package name and bundle id support to init command

## Possible Implementations
Add `--package` options to `init` command which allows the user to specify bundleid/packagename

### PR DESCRIPTION
**Pull request by TMisiukiewicz**
Summary:
---------
Closes #1269 
Relates to #1111 

Added `--package-name` option to `init` command, which allows initialising project with custom package name for Android and iOS.

Test Plan:
----------
Use `init` command with `--package-name` option:
1. If package name is incorrect e.g. it contains any special chars except dots or have only one segment it should end up with error. Examples: `example`, `com.example@2.app`
2. Initialise project with `--package-name example.app` - project should be initialised with `com.example.app` Android package name and iOS bundleID
3. Initialise project with `--package-name com.example.app` - project should be initialised with `com.example.app` Android package name and iOS bundleID

For 2. and 3., app should be installed on the device/emulator named by project name - not package name.

### CODE REVIEW

**Comment by thymikee** [comment]
Can we have some tests for this?

File: packages/cli/src/commands/init/index.ts [hunk 1]
```diff
@@ -32,5 +32,10 @@ export default {
       name: '--skip-install',
       description: 'Skips dependencies installation step',
     },
+    {
+      name: '--package-name <string>',
+      description:
+        'Inits a project with a custom package name (Android) and bundle ID (iOS), e.g. com.example.app',
+    },
```
Comment (szymonrybczak) [inline packages/cli/src/commands/init/index.ts:39]: Let's add this to docs 😄

**Review by szymonrybczak — APPROVED** [review]
LGTM! 🚀

File: packages/cli/src/commands/init/index.ts (hunk 1 above)
Comment (TMisiukiewicz) [reply packages/cli/src/commands/init/index.ts:39]: done ✅

**Comment by TMisiukiewicz** [comment]
> Can we have some tests for this?

tests added ✅

**Review by adamTrz — APPROVED** [review]
🚀

File: packages/cli/src/commands/init/editTemplate.ts [hunk 2]
```diff
@@ -83,32 +101,180 @@ async function processDotfiles(filePath: string) {
...
+
+      for (const segment of segmentsList) {
+        fs.mkdirSync(segment);
+        process.chdir(segment);
+      }
+
+      fs.rename(
+        `${filePath}/${segmentsList[segmentsList.length - 1]}`,
+        process.cwd(),
+      );
```
Comment (thymikee) [inline packages/cli/src/commands/init/editTemplate.ts:147]: Those renames can fail and `createAndroidPackagePaths` also isn't guarded by try/catch, so the user would be left blind when `fs.rename` throws. Please make sure to guard this case

File: packages/cli/src/commands/init/editTemplate.ts [hunk 3]
```diff
@@ -83,32 +101,180 @@ async function processDotfiles(filePath: string) {
...
+  }
+}
+
+export async function replacePlaceholderWithPackageName({
   projectName,
   placeholderName,
-  placeholderTitle = DEFAULT_TITLE_PLACEHOLDER,
-  projectTitle = projectName,
-}: PlaceholderConfig) {
-  logger.debug(`Changing ${placeholderName} for ${projectName} in template`);
```
Comment (thymikee) [inline packages/cli/src/commands/init/editTemplate.ts:92]: why is this removed?

**Comment by bryanltobing** [comment]
can we have an example for this?

I tried this 

```bash
npx react-native@latest init [APP_NAME] --package-name [PACKAGE_NAME] --title [APP_TITLE]
```

but it doesn't work

**Comment by szymonrybczak** [comment]
Hey @bryanltobing, what error you get if you try this command? And also could you please provide output from `npx react-native@latest -v` just to make sure that you're working on correct version 😅

**Comment by bryanltobing** [comment]
it does work now. sorry for the inconvenient

**Comment by javier8340** [comment]
packages only can start by "com"? can be changed as user sets? for example if domain start with "org" or by country domain

**Comment by TMisiukiewicz** [comment]
hi @javier8340, it was fixed in #1915
